Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,601 @@
+
+    m_mimeType = makeString(m_hasVideo ? "video/webm"_s : "audio/webm"_s, "; codecs=\""_s, m_hasVideo ? codecStringForMediaVideoCodecId(m_videoCodec) : ""_s, m_hasVideo && m_hasAudio ? ","_s : ""_s, m_hasAudio ? codecStringForMediaVideoCodecId(m_audioCodec) : ""_s, "\""_s);
+    return true;
+}
+
+void MediaRecorderPrivateWriterWebM::enqueueCompressedAudioSampleBuffers()
+{
+    ASSERT(m_hasAudio);
+
+    if (!m_audioCompressor || m_audioCompressor->isEmpty()) {
+        LOG(MediaStream, "enqueueCompressedAudioSampleBuffers: empty queue");
+        return;
+    }
+
+    if (!m_audioFormatDescription) {
+        m_audioFormatDescription = PAL::CMSampleBufferGetFormatDescription(RefPtr { m_audioCompressor }->getOutputSampleBuffer());
+
+        const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
+        if (!asbd)
+            return;
+
+        m_audioTrackIndex = m_delegate->addAudioTrack(*asbd, mkvCodeIcForMediaVideoCodecId(m_audioCodec));
+        if (!m_audioTrackIndex)
+            return;
+
+        maybeStartWriting();
+    }
+
+    if (!m_hasStartedWriting)
+        return;
+
+    while (RetainPtr sampleBlock = RefPtr { m_audioCompressor }->takeOutputSampleBuffer()) {
+        PAL::CMSampleBufferCallBlockForEachSample(sampleBlock.get(), [&] (CMSampleBufferRef sampleBuffer, CMItemCount) -> OSStatus {
+            assertIsCurrent(MainThreadDispatcher::singleton());
+            m_encodedAudioFrames.append(sampleBuffer);
+            LOG(MediaStream, "enqueueCompressedAudioSampleBuffers:Receiving compressed audio frame: queue:%zu first:%f last:%f", m_encodedAudioFrames.size(), sampleTime(m_encodedAudioFrames.first()).value(), sampleTime(m_encodedAudioFrames.last()).value());
+            return noErr;
+        });
+    }
+}
+
+void MediaRecorderPrivateWriterWebM::maybeStartWriting()
+{
+    m_hasStartedWriting = (!m_hasAudio || m_audioFormatDescription) && (!m_hasVideo || m_videoEncoder);
+}
+
+void MediaRecorderPrivateWriterWebM::appendVideoFrame(VideoFrame& frame)
+{
+    if (m_isStopping)
+        return;
+
+    if (!m_firstVideoFrameProcessed) {
+        m_firstVideoFrameProcessed = true;
+        VideoEncoder::Config config { static_cast<uint64_t>(frame.presentationSize().width()), static_cast<uint64_t>(frame.presentationSize().height()), false, m_videoBitsPerSecond };
+
+        m_videoTrackIndex = m_delegate->addVideoTrack(config.width, config.height, mkvCodeIcForMediaVideoCodecId(m_videoCodec));
+
+        VideoEncoder::create(codecStringForMediaVideoCodecId(m_videoCodec), config, [weakThis = ThreadSafeWeakPtr { *this }, this](auto&& result) {
+            assertIsMainThread();
+            RefPtr protectedThis = weakThis.get();
+            if (!result || !protectedThis)
+                return;
+            m_videoEncoder = result.value().moveToUniquePtr();
+            m_videoEncoder->setRates(m_videoBitsPerSecond, 0);
```

Comment:
Should we call `encodePendingVideoFrames ` here?

Reply:
it is called, just after the call to `VideoEncoder::create(` ; but I can see how it might be preferable to do so here too.

Reply:
There is no m_videoEncoder yet after so encodePendingVideoFrames() below will be a no-op.